README.md
# pass-client

[![GitHub stars](https://img.shields.io/github/stars/clayrisser/pass-client.svg?style=social&label=Stars)](https://github.com/clayrisser/pass-client)

> nodejs client for the standard unix password manager

Please ★ this repo if you found it useful ★ ★ ★

## Features

- list passwords
- find a password
- edit a password

## Installation

```sh
npm install --save pass-client
```

## Dependencies

- [NodeJS](https://nodejs.org)

## Usage

[Contribute](https://github.com/clayrisser/pass-client/blob/master/CONTRIBUTING.md) usage docs

## Support

Submit an [issue](https://github.com/clayrisser/pass-client/issues/new)

## Screenshots

[Contribute](https://github.com/clayrisser/pass-client/blob/master/CONTRIBUTING.md) a screenshot

## Contributing

Review the [guidelines for contributing](https://github.com/clayrisser/pass-client/blob/master/CONTRIBUTING.md)

## License

[Apache-2.0 License](https://github.com/clayrisser/pass-client/blob/master/LICENSE)

[Clay Risser](https://clayrisser.com) © 2021

## Changelog

Review the [changelog](https://github.com/clayrisser/pass-client/blob/master/CHANGELOG.md)

## Credits

- [Clay Risser](https://clayrisser.com) - Author

## Support on Liberapay

A ridiculous amount of coffee ☕ ☕ ☕ was consumed in the process of building this project.

[Add some fuel](https://liberapay.com/clayrisser/donate) if you'd like to keep me going!

[![Liberapay receiving](https://img.shields.io/liberapay/receives/clayrisser.svg?style=flat-square)](https://liberapay.com/clayrisser/donate)
[![Liberapay patrons](https://img.shields.io/liberapay/patrons/clayrisser.svg?style=flat-square)](https://liberapay.com/clayrisser/donate)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.0.1] - 2021-07-05

### Added

- Initial release


src/pass.spec.ts
/*
 * File: /src/pass.spec.ts
 * Project: pass-client
 * File Created: 05-07-2021 14:35:37
 * Author: Clay Risser <email@clayrisser.com>
 * -----
 * Last Modified: 05-07-2021 15:25:28
 * Modified By: Clay Risser <email@clayrisser.com>
 * -----
 * Silicon Hills LLC (c) Copyright 2021
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Pass from '~/pass';

describe('new Pass().list()', () => {
  it('should list passwords with browserpass', async () => {
    const pass = new Pass();
    const data = await pass.list();
    expect(data).toMatchObject({
      files: {}
    });
    expect(typeof data.files.default?.length).toBe('number');
  });

  it('should list passwords without browserpass', async () => {
    const pass = new Pass({ browserpass: false });
    const data = await pass.list();
    expect(data).toMatchObject({
      files: {}
    });
    expect(typeof data.files.default?.length).toBe('number');
  });

  it('with and without browserpass should match', async () => {
    const withBrowserpassPass = new Pass();
    const withoutBrowserpassPass = new Pass({ browserpass: false });
    expect((await withBrowserpassPass.list()).files.default?.sort()).toEqual(
      (await withoutBrowserpassPass.list()).files.default?.sort()
    );
  });
});


src/bin.ts
/*
 * File: /src/bin.ts
 * Project: pass-client
 * File Created: 05-07-2021 15:25:51
 * Author: Clay Risser <email@clayrisser.com>
 * -----
 * Last Modified: 05-07-2021 15:55:37
 * Modified By: Clay Risser <email@clayrisser.com>
 * -----
 * Silicon Hills LLC (c) Copyright 2021
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Pass } from './index';

const pass = new Pass();

const logger = console;

(async () => {
  switch (process.argv[2]) {
    case 'show': {
      const { contents } = await pass.show(process.argv[3]);
      process.stdout.write(contents);
      break;
    }
    case 'insert': {
      await pass.insert(process.argv[3], process.argv[4]);
      break;
    }
  }
})().catch(logger.error);


src/index.ts
/*
 * File: /src/index.ts
 * Project: pass-client
 * File Created: 05-07-2021 14:11:03
 * Author: Clay Risser <email@clayrisser.com>
 * -----
 * Last Modified: 05-07-2021 16:06:29
 * Modified By: Clay Risser <email@clayrisser.com>
 * -----
 * Silicon Hills LLC (c) Copyright 2021
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import NativeMessage from './nativeMessage';
import Pass from './pass';

export { NativeMessage, Pass };

export * from './nativeMessage';
export * from './pass';
export * from './types';


src/types.ts
/*
 * File: /src/types.ts
 * Project: pass-client
 * File Created: 05-07-2021 14:22:57
 * Author: Clay Risser <email@clayrisser.com>
 * -----
 * Last Modified: 05-07-2021 14:22:59
 * Modified By: Clay Risser <email@clayrisser.com>
 * -----
 * Silicon Hills LLC (c) Copyright 2021
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export interface HashMap<T = any> {
  [key: string]: T;
}

export type DeepHashMap<T = any> = HashMap<T | DeepHashMap<T>>;

export interface StoreSettings {
  gpgPath: string;
}

export interface Store {
  id: string;
  name: string;
  path: string;
  settings: StoreSettings;
}

export interface Settings {
  gpgPath: string;
  stores: HashMap<Store>;
}

export interface Request {
  action: Action;
  echoResponse: any;
  file: string;
  settings: Settings;
  storeId: string;
}

export interface Response<D = DeepHashMap<any>> {
  status: Status;
  version: number;
  data: D;
  code?: any;
  params?: any;
}

export interface ConfigureResponse {
  defaultStore: Store;
  storeSettings: HashMap<string>;
}

export enum Action {
  configure = 'configure',
  echo = 'echo',
  fetch = 'fetch',
  list = 'list'
}

export enum Status {
  ok = 'ok'
}


src/nativeMessage.ts
/*
 * File: /src/nativeMessage.ts
 * Project: pass-client
 * File Created: 05-07-2021 14:20:58
 * Author: Clay Risser <email@clayrisser.com>
 * -----
 * Last Modified: 05-07-2021 15:55:18
 * Modified By: Clay Risser <email@clayrisser.com>
 * -----
 * Silicon Hills LLC (c) Copyright 2021
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import execa from 'execa';
import { Request } from '~/types';

export default class NativeMessage {
  constructor(private binary: string) {}

  async request<R = any>(request: Request): Promise<R> {
    const buffer = this.contentToBytes(JSON.stringify(request));
    let result: any = {};
    try {
      result = JSON.parse(
        Buffer.from(
          (
            await execa(this.binary, [], {
              input: buffer
            })
          ).stdout
        )
          .subarray(4)
          .toString()
      );
    } catch (err) {
      if (
        !/Unexpected\stoken\s.\sin\sJSON\sat\sposition\s\d/.test(err.message)
      ) {
        throw err;
      }
      result = JSON.parse(
        Buffer.from(
          (
            await execa(this.binary, [], {
              input: buffer
            })
          ).stdout
        )
          .subarray(6)
          .toString()
      );
    }
    return result;
  }

  private contentToBytes(content: string): Buffer {
    return Buffer.concat([
      this.lengthToBytes(content.length),
      Buffer.from(content)
    ]);
  }

  private lengthToBytes(length: number) {
    const buffer = Buffer.alloc(4);
    let hex = '';
    Array.from(length.toString(16).padStart(8, '0')).forEach(
      (hexChar: string, i: number) => {
        if (i % 2 === 0) {
          hex = hexChar;
        } else {
          hex += hexChar;
          buffer[(i + 1) / 2 - 1] = parseInt(hex, 16);
          hex = '';
        }
      }
    );
    return buffer.reverse();
  }
}

export interface NativeMessageOptions {}


src/pass.ts
/*
 * File: /src/pass.ts
 * Project: pass-client
 * File Created: 05-07-2021 14:25:01
 * Author: Clay Risser <email@clayrisser.com>
 * -----
 * Last Modified: 05-07-2021 15:49:21
 * Modified By: Clay Risser <email@clayrisser.com>
 * -----
 * Silicon Hills LLC (c) Copyright 2021
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import execa from 'execa';
import os from 'os';
import path from 'path';
import which from 'which';
import NativeMessage from '~/nativeMessage';
import { Action, HashMap, Response } from '~/types';

export default class Pass {
  private options: PassOptions;

  constructor(options: PassOptions = {}) {
    this.options = {
      browserpass: true,
      ...options
    };
  }

  private _browserpass: NativeMessage | null = null;

  private _browserpassPath = '';

  private _gpgPath = '';

  private _passPath = '';

  private _findPath = '';

  private _opensslPath = '';

  async list(): Promise<ListData> {
    const browserpass = await this.getBrowserpass();
    if (browserpass) {
      return (
        await browserpass.request<Response<ListData>>({
          action: Action.list,
          echoResponse: '',
          file: '',
          storeId: 'default',
          settings: {
            gpgPath: await this.getGpgPath(),
            stores: {
              default: {
                id: 'default',
                name: 'default',
                path: this.passwordStorePath,
                settings: {
                  gpgPath: await this.getGpgPath()
                }
              }
            }
          }
        })
      ).data;
    }
    const files = (
      await execa(
        await this.getFindPath(),
        ['.', '-type', 'f', '-name', '*.gpg'],
        {
          cwd: this.passwordStorePath
        }
      )
    ).stdout
      .split('\n')
      .map((file: string) => file.substr(2));
    return { files: { default: files } };
  }

  async show(file: string): Promise<FetchData> {
    if (!/.gpg$/.test(file)) file = `${file}.gpg`;
    const browserpass = await this.getBrowserpass();
    if (browserpass) {
      return (
        await browserpass.request<Response<FetchData>>({
          action: Action.fetch,
          echoResponse: '',
          file,
          storeId: 'default',
          settings: {
            gpgPath: await this.getGpgPath(),
            stores: {
              default: {
                id: 'default',
                name: 'default',
                path: this.passwordStorePath,
                settings: {
                  gpgPath: await this.getGpgPath()
                }
              }
            }
          }
        })
      ).data;
    }
    const contents = (
      await execa(await this.getPassPath(), [
        'show',
        file.replace(/.gpg$/g, '')
      ])
    ).stdout;
    return { contents };
  }

  async insert(file: string, content: string): Promise<void> {
    await execa(
      `(echo "${Buffer.from(content).toString(
        'base64'
      )}") | ${await this.getOpensslPath()} base64 -d | pass insert -m ${file.replace(
        /.gpg$/g,
        ''
      )}`,
      { shell: true }
    );
  }

  private async getGpgPath(): Promise<string> {
    if (this._gpgPath.length) {
      return this._gpgPath;
    }
    this._gpgPath = this.options.gpgPath || (await which('gpg')) || 'gpg';
    return this._gpgPath;
  }

  private async getPassPath(): Promise<string> {
    if (this._passPath.length) {
      return this._passPath;
    }
    this._passPath = this.options.passPath || (await which('pass')) || 'pass';
    return this._passPath;
  }

  private async getFindPath(): Promise<string> {
    if (this._findPath.length) {
      return this._findPath;
    }
    this._findPath = this.options.findPath || (await which('find')) || 'find';
    return this._findPath;
  }

  private async getOpensslPath(): Promise<string> {
    if (this._opensslPath.length) {
      return this._opensslPath;
    }
    this._opensslPath =
      this.options.opensslPath || (await which('openssl')) || 'openssl';
    return this._opensslPath;
  }

  private async getBrowserpassPath(): Promise<string | null> {
    if (this._browserpassPath.length) {
      return this._browserpassPath;
    }
    const browserpassPath =
      this.options.browserpassPath || (await which('browserpass'));
    if (!browserpassPath) return null;
    this._browserpassPath = browserpassPath;
    return this._browserpassPath;
  }

  private async getBrowserpass(): Promise<NativeMessage | null> {
    if (!this.options.browserpass) return null;
    if (this._browserpass) return this._browserpass;
    const browserpassPath = await this.getBrowserpassPath();
    if (!browserpassPath) return null;
    this._browserpass = new NativeMessage(browserpassPath);
    return this._browserpass;
  }

  private get passwordStorePath() {
    return this.options.passwordStorePath
      ? path.resolve(this.options.passwordStorePath)
      : path.resolve(os.homedir(), '.password-store');
  }
}

export interface ListData {
  files: HashMap<string[]>;
}

export interface FetchData {
  contents: string;
}

export interface PassOptions {
  browserpass?: boolean;
  browserpassPath?: string;
  opensslPath?: string;
  findPath?: string;
  gpgPath?: string;
  passPath?: string;
  passwordStorePath?: string;
}
